metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteReferencesService.java

```diff
@@ -0,0 +1,177 @@
+package com.linkedin.metadata.entity;
+
+import com.linkedin.data.DataComplex;
+import com.linkedin.data.DataList;
+import com.linkedin.data.DataMap;
+import com.linkedin.data.schema.ArrayDataSchema;
+import com.linkedin.data.schema.DataSchema;
+import com.linkedin.data.schema.PathSpec;
+import com.linkedin.data.schema.RecordDataSchema;
+import com.linkedin.data.template.RecordTemplate;
+import com.linkedin.entity.Aspect;
+import java.util.List;
+import java.util.ListIterator;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Utility class that encapsulates the logic on how to modify a {@link RecordTemplate} in place to remove a single value
+ * following a concrete set of rules.
+ *
+ * It does this by a recursive tree traversal method, based on an aspect path provided for a given aspect.
+ * This so that we don't have to scan the entire aspect for the value to be removed and then figure out how to apply
+ * logical rules based on upstream optionality definitions.
+ *
+ * For more information see {@link #getAspectWithReferenceRemoved(String, RecordTemplate, DataSchema, PathSpec)}
+ */
+@Slf4j
+public class DeleteReferencesService {
+
+  private DeleteReferencesService() { }
+
+  /**
+   *
+   * Utility method that provides utility methods to remove fields from a given aspect based on it's aspect spec that
+   * follows the following logic:
+   *
+   * 1. If field is optional and not part of an array → remove the field.
+   * 2. If is a field that is part of an array (has an `*` in the path spec)
+   *  → go up to the nearest array and remove the element.
+   *  Extra → If array only has 1 element which is being deleted→ optional rules (if optional set null, otherwise delete)
+   * 3. If field is non-optional and does not belong to an array delete if and only if aspect becomes empty.
```

Comment:
If field is non-optional and does not belong to an array delete if and only if aspect becomes empty. 

"If and only if aspect becomes empty". --> What does this part mean?

Reply:
That if all other fields in the aspect are optional and are set to null, then deleting the field is equivalent to deleting the aspect, otherwise the aspect would be invalid as information would be lost.

Reply:
Got it. So we do issue a DELETE on the aspect?